File: javascript/perfect-numbers/perfect-numbers.js
// https://www.nayuki.io/res/calculate-divisors-javascript/calculate-divisors.js
function listDivisors(n) {
  if (n < 1) {
    throw new Error('Argument error');
  }
  // Edge case, easier to deal with here
  if (n === 1) {
    return [];
  }
  const small = [];
  const large = [];
  const end = Math.floor(Math.sqrt(n));
  for (let i = 1; i <= end; i += 1) {
    if (n % i === 0) {
      small.push(i);
      // Don't include a square root twice or n
      if (i * i !== n) {
        large.push(n / i);
      }
    }
  }
  // Remove the number itself
  const largeWithoutN = large.slice(1);
  // This is unnecessary, since we don't care about the order.
  // largeWithoutN.reverse();
  return small.concat(largeWithoutN);
}

function sum(arr) {
  return arr.reduce((a, b) => a + b, 0);
}

function PerfectNumbers() {
  this.classify = function classify(n) {
    if (n <= 0) {
      return 'Classification is only possible for natural numbers.';
    }
    const divisorsSum = sum(listDivisors(n));
    if (divisorsSum > n) {
      return 'abundant';
    } else if (divisorsSum < n) {
      return 'deficient';
    }
    return 'perfect';
  };
}

module.exports = PerfectNumbers;

File: haskell/perfect-numbers/src/PerfectNumbers.hs
module PerfectNumbers (classify, Classification(..)) where

import Data.Numbers.Primes (primeFactors)
import Data.List (tails, group)

data Classification = Deficient | Perfect | Abundant deriving (Eq, Show)

classify :: Int -> Maybe Classification
classify i = classify' <$> aliquot i
  where classify' a
          | a > i     = Abundant
          | a < i     = Deficient
          | otherwise = Perfect

aliquot :: Int -> Maybe Int
aliquot = go
  where go i = sum <$> divisors i

divisors :: Int -> Maybe [Int]
divisors x
  | x <  1 = Nothing
  | x == 1 = Just []
  | otherwise = let ps      = primeFactors x
                    pgroups = group ps
                in Just . tail . map product . partitions $ pgroups

partitions :: [[a]] -> [[a]]
partitions [] = [[]]
partitions (xs:xss) = let rest = partitions xss
                      in concatMap (\xs' -> map (xs' ++) rest) $ tails xs